base/lock.jl

```diff
@@ -342,40 +346,62 @@ end
 
 
 """
-    Event()
+    Event([autoreset=false])
 
 Create a level-triggered event source. Tasks that call [`wait`](@ref) on an
```

Comment:
I wonder if we can use a different name though? It's conceptually much cleaner if we can say `Condtion` is edge-triggered and `Event` is edge-triggered. We can still share most of the implementation by using `::Union{Event,AutoreSetThingy}`. For example, would it be crazy to call this a `Signal`? Or, if we go with `Event([autoreset=false])`, I think we need to at least reword the dosctring "Create a level-triggered event source."

Reply:
Being level-triggered doesn't preclude being resettable, though I agree the historic choices of names was unfortunate, I don't think we should attempt to deviate from them now.

Reply:
My point was the opposite. I was not suggesting moving away from the historical naming.

Rather, one of my suggestions was to introduce a new type for `Event(true)` and keep the old semantics of `Event()` as-is.

Reply:
This doesn't change the semantics of `Event()`, but extends it with a useful feature (the ability to reset it manually/automatically) common to other uses of the name (in kevent, the name of this flag is EV_CLEAR, in Event, the name of this flag is bManualReset)

Reply:
Hmm..., I think it changes how much you can infer locally about the correctness of the code. IIUC, with `autoreset = false`, I can conclude that

```julia
ev::Event
ref::Ref
@sync for i in 1:2
    @spawn begin
        wait(ev)
        ref[] = i
    end
end
```

has a data race. With `autoreset = true`, it may not have a data race (depending on the use of `notify`). This is obviously not the best example. But, I think it at least shows that this PR does change the semantics of `Event`.

I can't think of any reasons for avoiding different types to represent `autoreset`-or-not other than coming up a good name is hard (which *can* also be reflected to the discoverability of the API; however, I'd assume this is negligible as long as the new API is linked from `Event` docstring). It's hard to come up with a decent name but that's a one time cost. Readability of the code has an effect to the cost proportional to the time and the number of Julia programmers using it.

Reply:
This feature seems OK to me. `Event` is still level-triggered; the only difference is that the level might not last forever. It should be clear from context which events might change status, e.g. "data available" vs. " task has exited".

Reply:
Is there a downside in providing the information on autoreset-or-not as a type in the code (e.g., `Event` vs `AutoresetEvent`)? Concurrent programming is especially tricky and so more information (which is often locally obvious) seems to be strictly better to me.

That said, I don't have enough experience in using autoreset-`Event` and so I don't know if the upside I'm arguing is negligible. Also, users can always create a wrapper type `AutoresetEvent` that ensures `Event(true)` is called in the constructor if they really need it (but this makes reading others' code harder).